OJ-990: Verify the signature returned by vc

jkunle opened the pull request:
## Proposed changes

This PR is to verify that the SignedJWT by ECDSA algorithm returned is in the expected concat format
The verification keys are configured here https://github.com/alphagov/di-ipv-config/pull/958

### What changed

This as result of https://github.com/alphagov/di-ipv-cri-lib/pull/64 which makes the VC return compliant SignedJWT format according to [RFC 7518, section 3.4: Digital Signature with ECDSA](https://www.rfc-editor.org/rfc/rfc7518#section-3.4)

### Why did it change

SignedJWT formally returned DER format which is non-compliant with respect to the above named specification. 

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [OJ-990](https://govukverify.atlassian.net/browse/OJ-990)

## Checklists

see: https://github.com/alphagov/di-ipv-cri-lib/pull/64 is merged

jkunle committed:
OJ-990: Verify the signature returned by vc

tal-nagra commented on di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/utils/HandlerHelper.java:
```diff
@@ -393,4 +397,19 @@ private RSAKey getEncryptionPublicKey(CredentialIssuer credentialIssuer)
                         Base64.getDecoder().decode(credentialIssuer.publicEncryptionJwkBase64())));
     }
+
+    public boolean isSignedUsingECDSA(SignedJWT signedJWT) throws JOSEException {
```
To adhere to the single responsibility principle this method should be split into 2 methods:
- `void checkSignatureFormat(SignedJWT signedJWT)` this would contain the code on lines 402-406
- `boolean verifySignedJwt(SignedJWT signedJWT)` this would contain the code on lines 408-413

tal-nagra commented on di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/CoreStubConfig.java:
```diff
@@ -42,6 +42,9 @@ public class CoreStubConfig {
     public static final String CORE_STUB_SIGNING_PRIVATE_KEY_JWK_BASE64 =
             getConfigValue("CORE_STUB_SIGNING_PRIVATE_KEY_JWK_BASE64", null);
+
+    public static final String CORE_STUB_SIGNING_PUBLIC_JWK =
```
Does the README or any other files need to be updated to set the environment variable `CORE_STUB_SIGNING_PUBLIC_JWK` ?

jkunle replied:
The value comes from di-ipv-config see `https://github.com/alphagov/di-ipv-config/blob/main/stubs/di-ipv-core-stub/.env`

tal-nagra commented on di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/handlers/CoreStubHandler.java:
```diff
@@ -147,12 +147,18 @@ private void setQuestions() {
-                                .toString();
+                                handlerHelper.getUserInfo(accessToken, credentialIssuer, state));
+
+                if (handlerHelper.isSignedUsingECDSA(signedJWT)) {
```
As per the comments below invoke:
1) handlerHelper.checkSignatureFormat(signedJWT);`
2) `if (!handlerHelper.verifySignedJwt(signedJWT) { throw ... }`

jkunle committed:
OJ-990: Add VC signing property to CredentialIssuer

Add publicVCSigningVerificationJwkBase64 when VC signs the JWT it produces
It can be verified the the 64 encode VC signing JWT held by
publicVCSigningVerificationJwkBase64

publicVCSigningVerificationJwkBase64 is set in the cris-local-dev.yaml and
cris-dev.yaml file in di-ipv-config

jkunle committed:
OJ-990: Make verify signature concat format optional

This is the ensure backward compatibility

idrop commented on di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/utils/HandlerHelper.java:
```diff
@@ -393,4 +397,21 @@ private RSAKey getEncryptionPublicKey(CredentialIssuer credentialIssuer)
                         Base64.getDecoder().decode(credentialIssuer.publicEncryptionJwkBase64())));
     }
+
+    public boolean checkSignatureFormat(SignedJWT signedJWT) throws JOSEException {
```
can we rename this to indicate that its checking for a [ES256](https://ldapwiki.com/wiki/ES256) signing algorithm
